pkg/dynamicrestmapper/dynamicrestmapper.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2025 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamicrestmapper
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type DynamicRESTMapper struct {
```

Comment:
not sure where this is going. The data structures below are just the `DefaultRESTMapper` ones, right?

What we need is that for every logical cluster in the system, i.e. something like `map[logicalcluster.Name]DefaultRESTMapper`.

Reply:
As discussed async, this needs to comply with the DefaultRestMapper interface but be logicalcluster compabile so every process in the shard could resolve particular cluster mappings. And as inspiration index https://github.com/kcp-dev/kcp/blob/main/pkg/index/index.go#L79 code could be referred. 

Basically what sttts said :)

Reply:
I was thinking something under lines:
```
m := mapper.For(logicalcluster) 
or 
mapper.For(logicalcluster). KindFor()
```
which is basically same. Not sure if we want "global restmapper per shard at this point".

Reply:
Thanks for the examples! It would be great if we could use `DefaultRESTMapper` as-is, but if I understood correctly, we'll to delete the mappings too (when a CRD is removed), and since `DefaultRESTMapper` doesn't implement deletion, we'd need to copy-paste all its code first :/

Reply:
>  "global restmapper per shard at this point".

This is only for the local shard.

> and since DefaultRESTMapper doesn't implement deletion, we'd need to copy-paste all its code first :/

The existing code is pretty complex. Wondering how we can soft-fork it. I could see that with one additional constructor and some wrapping with locking we could reuse all the existing code, while being able to change the involved maps.

Reply:
Thinking out loud:

soft-fork = maybe copy over through go generate and add a removal method in another file.